README.md

```diff
@@ -1,10 +1,239 @@
+    var_dump($jsonArray);
+
+    echo "This text is displayed AFTER Json response\n";
+}
+```
+Like with the `yield` keyword,
+the [`EventLoop::wait`](src/EventLoop.php) method will return the resolved value of the input [`Promise`](src/Promise.php),
+but remember that you should use it only once during execution.
+
+### Concurrency
+To reveal the true power of asynchronous programming, we have to introduce *concurrency* in our program.
+If our goal is to send only one HTTP request and wait for it,
+there is no gain to deal with an asynchronous request.
+However, as soon as you have at least two goals to reach,
+asynchronous functions will improve your performances thanks to concurrency.
+To resolve several independent [`Promises`](src/Promise.php),
+use [`EventLoop::promiseAll`](src/EventLoop.php) method to create a new [`Promise`](src/Promise.php)
+that will be resolved when all others are resolved.
+```php
+<?php
+
+use M6Web\Tornado\EventLoop;
+use M6Web\Tornado\HttpClient;
+use Psr\Http\Message\RequestInterface;
+
+function mySynchronousAction(EventLoop $eventLoop, HttpClient $httpClient, array $requests)
+{
+    $myPromises = [];
+    /** @var RequestInterface $request */
+    foreach ($requests as $request) {
+        $myPromises[] = getJsonResponse($httpClient, $request);
+    }
+    
+    $allJsonArrays = $eventLoop->wait(
+        $eventLoop->promiseAll(...$myPromises)
+    );
+    
+    var_dump($allJsonArrays);
+}
+```
+
+It's important to note that 
+it will be more efficient to use [`EventLoop::promiseAll`](src/EventLoop.php)
+instead of waiting each input [`Promise`](src/Promise.php) consecutively,
+because of concurrency.
+Each time that you have several promises to resolve,
+ask yourself if you could wait them concurrently, especially when you deal with `foreach` loops. 
+
+### Resolving your own promises
+By design, you cannot resolve a promise by yourself, you will need a [`Deferred`](src/Deferred.php).
+It allows you to create a [`Promise`](src/Promise.php) and to resolve (or reject) it
+while not exposing these advanced controls.
+```php
+<?php
+
+use M6Web\Tornado\Deferred;
+use M6Web\Tornado\EventLoop;
+use M6Web\Tornado\HttpClient;
+use M6Web\Tornado\Promise;
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\ResponseInterface;
+
+function loopUntilStatus418(EventLoop $eventLoop, HttpClient $httpClient, RequestInterface $request, Deferred $deferred)
+{
+    do {
+        yield $eventLoop->delay(1000 /*ms*/);
+        /** @var ResponseInterface $response */
+        $response = yield $httpClient->sendRequest($request);
+    } while ($response->getStatusCode() !== 418);
+
+    $deferred->resolve($response->getBody());
+}
+
+function get418Response(EventLoop $eventLoop, HttpClient $httpClient, RequestInterface $request): Promise
+{
+    $deferred = $eventLoop->deferred();
+    $eventLoop->async(loopUntilStatus418($eventLoop, $httpClient, $request, $deferred));
```

Comment:
are you forced to modify the deferred by reference in your example? can't you return it from async? 🤔

Reply:
`loopUntilStatus418` is an asynchronous function, so if I return a deferred, it will be a promise of a deferred… #mindblown It's not the expected behavior 😁 
I agree that this use case is not perfect, because we could use directly the `Promise` returned by `async`… I will think about it, but I am pretty sure that I will be obliged to pass deferred by reference, since it's the only way to pass informations between asynchronous functions… 🤔